src/cargo/ops/cargo_rustc/job_queue.rs

```diff
@@ -113,56 +113,100 @@ impl<'a> JobQueue<'a> {
     pub fn execute(&mut self, cx: &mut Context) -> CargoResult<()> {
         let _p = profile::start("executing the job graph");
 
+        // We need to give a handle to the send half of our message queue to the
+        // jobserver helper thrad. Unfortunately though we need the handle to be
+        // `'static` as that's typically what's required when spawning a
+        // thread!
+        //
+        // To work around this we transmute the `Sender` to a static lifetime.
+        // we're only sending "longer living" messages and we should also
+        // destroy all references to the channel before this function exits as
+        // the destructor for the `helper` object will ensure the associated
+        // thread i sno longer running.
```

Comment:
That's what I originally thought yeah but actually I don't think so (unfortunately). I believe `Drop for Sender` may run destructors for items in the internal queue (not received yet), which means that if you persist a `Sender` beyond the lifetime of the item I think it'll access data outside of its lifetime.

(not in this case though, the `Sender` should always go away with the stack frame.